File: test/csharp/SpecGateFixtures/Conformance/Stateful/VoidOperation/void_operation.cs
using SpecGate.Annotations;
namespace SpecGateFixtures.Conformance.Stateful.VoidOperation;

/// <summary>
/// Void operation: <c>log</c> returns nothing but mutates state and echoes an
/// input, verifying capture of inputs and state changes for unit-returning ops.
/// </summary>
public class Logger
{
    /// <summary>The number of messages logged; incremented on each call.</summary>
    [SpecEvent("count")]
    public int Count { get; set; }

    /// <summary>Builds a logger with a zero count.</summary>
    /// <returns>A new <see cref="Logger"/> with <see cref="Count"/> 0.</returns>
    [SpecSetup("log")]
    public static Logger Make() => new() { Count = 0 };

    /// <summary>Records a message, incrementing the count (no return value).</summary>
    /// <param name="msg">The message to log (spec input <c>msg</c>).</param>
    [SpecOperation("log")]
    public void Log([SpecInput("msg")] string msg) => Count += 1;
}

File: test/rust/crates/specgate-fixtures/src/conformance/stateful/void_operation.rs
// Operation with no return value (unit type).
use specgate::*;

#[spec_setup("log")]
pub fn make_logger() -> Logger {
    Logger { count: 0 }
}

#[derive(SpecEvent)]
pub struct Logger {
    #[spec_event]
    pub count: i32,
}

impl Logger {
    #[spec_operation("log", spec = "fixture.void_operation")]
    pub fn log(&mut self, #[spec_input("msg")] _msg: &str) {
        self.count += 1;
    }
}
